Feature/access request gsi 1289

SilverLinings89 opened the pull request:
Still a WIP but making good progress now. Includes work on GSI-1300, GSI-1299 and GSI-1289.

Cito commented on src/app/data-access/services/data-access.service.ts:
```diff
@@ -0,0 +1,180 @@
+    () => this.#accessRequests.value() ?? [],
+  );
+  grantedAccessRequests = computed(() =>
+    this.accessRequests().filter((ar: AccessRequest) => ar.status == 'allowed'),
```
also filter out those that are expired (`access_ends` < now)

Cito replied:
these still need to be filtered out

Cito commented on src/app/data-access/services/data-access.service.ts:
```diff
@@ -0,0 +1,180 @@
+    }
+  };
+
+  #accessRequests = rxResource<AccessRequest[], string | null>({
```
This should be called more specifically userAccessRequests, because it returns only the requests of the current user. We will also need a resource allAccessRequets that returns the requests of all users for the access request manager. I'm using this approach in the ivaService.

Currently the service loads the requests as soon as the user logs in. In the ivaService I added a method loadUserIvas that triggers the loading only upon explicit request.

Cito commented on src/app/data-access/services/data-access.service.ts:
```diff
@@ -0,0 +1,180 @@
+      iva_id: '',
+    },
+  ]);
+
```
In the tests, you can also reuse the data from src/mocks/data.ts.

Cito commented on src/app/metadata/features/dataset-details/dataset-details.component.html:
```diff
@@ -4,7 +4,12 @@
-      <a mat-raised-button><mat-icon>key</mat-icon>Request Access</a>
+      <button
+        mat-raised-button
+        (click)="accessService.showNewAccessRequestDialog(datasetDetails().accession)"
```
I would wrap this as a component method. The accessService can then also be a private variable.

Cito replied:
Later, the button should become a separate component that would also change depending on whether access has already been granted or the request is already pending. I will add a task for this.

Cito commented on src/app/metadata/features/dataset-summary/dataset-summary.component.html:
```diff
@@ -9,7 +9,12 @@
+      mat-raised-button
+      (click)="accessService.showNewAccessRequestDialog(hit().id_)"
+    >
+      <mat-icon>key</mat-icon>Request Access
```
see above

Cito commented on src/app/metadata/services/metadata-search.service.ts:
```diff
@@ -49,7 +49,17 @@ export class MetadataSearchService {
+    );
+    console.log(newMassQueryUrl);
+
+    return newMassQueryUrl;
```
this should be reverted before merging

Cito replied:
There are still two console.logs.

Cito commented on src/mocks/data.ts:
```diff
@@ -425,6 +425,85 @@ export const searchResults: SearchResults = {
+    user_id: 'j.doe@ghga.de',
+    dataset_id: 'GHGAD588887987',
+    full_user_name: 'Dr. John Doe',
+    email: 'j.jdoe@home.org',
```
```suggestion
    email: 'doe@home.org',
```

Cito commented on src/mocks/data.ts:
```diff
@@ -425,6 +425,85 @@ export const searchResults: SearchResults = {
+export const accessRequests = [
+  {
+    id: '62bcc452-a70b-47c1-9870-55da40d8e45f',
+    user_id: 'j.doe@ghga.de',
```
```suggestion
    user_id: 'doe@test.dev',
```

Cito replied:
The emails have been changed in the other mock data, we should adapt here as well.

We should also add one or more access requests for Jane Roe (roe@test.dev) that can be later shown in the access request manager.

Cito commented on src/mocks/data.ts:
```diff
@@ -425,6 +425,85 @@ export const searchResults: SearchResults = {
+
+  {
+    id: '4ef4ccac-6c0a-4be6-9637-b33925178cea',
+    user_id: 'j.doe@ghga.de',
```
also change here and below

Cito commented on src/mocks/responses.ts:
```diff
@@ -119,6 +120,27 @@ export const responses: { [endpoint: string]: ResponseValue } = {
+   */
+  // Specific dataset and user access requests
+  'GET /api/ars/access-requests?dataset_id=GHGAD588887987&*': getAccessRequests(
+    'j.doe@ghga.de',
```
```suggestion
    'doe@test.dev',
```

Cito commented on src/mocks/responses.ts:
```diff
@@ -119,6 +120,27 @@ export const responses: { [endpoint: string]: ResponseValue } = {
+  ),
+
+  // Specific dataset and user access requests
+  'GET /api/ars/access-requests?*': getAccessRequests('j.doe@ghga.de'),
```
```suggestion
  'GET /api/ars/access-requests?*': getAccessRequests('doe@test.dev'),
```

Cito reviewed (changes requested):
Thanks, this already covers a lot of ground. Actually maybe too much for one PR - better to create smaller, more focused PRs corresponding to the tasks on the board.

I have made some change requests and suggestions below.

SilverLinings89 commented:
> Thanks, this already covers a lot of ground. Actually maybe too much for one PR - better to create smaller, more focused PRs corresponding to the tasks on the board.
> 
> I have made some change requests and suggestions below.

Yes, definitely. Sorry for this - I know it's really bad form and hard to review. I will try to do this better in the future. I struggle to find a point where to stop. Did the UI components first and then I didn't want to create an intermediate service implementation so I went right on with that ... The other way around (service, then components) would be better.

SilverLinings89 commented on eslint.config.js:
```diff
@@ -114,7 +114,12 @@ export default [
+                      'metadata',
+                      'verification-addresses',
+                      'access-requests',
+                      'data-access',
```
I wasn't 100% sure that's what you meant.

SilverLinings89 commented on src/app/data-access/features/data-access-request-modal/data-access-request-modal.component.html:
```diff
@@ -0,0 +1,58 @@
+      <mat-datepicker-toggle matIconSuffix [for]="until_picker"></mat-datepicker-toggle>
+      <mat-datepicker #until_picker></mat-datepicker>
+    </mat-form-field>
+
```
I'll implement it to work that way. Almost there anyway.

SilverLinings89 replied:
Implemented a first version.

SilverLinings89 replied:
May be broken, need to sleep :)

SilverLinings89 commented on src/app/data-access/services/data-access.service.ts:
```diff
@@ -0,0 +1,180 @@
+import { ConfigService } from '@app/shared/services/config.service';
+import { NotificationService } from '@app/shared/services/notification.service';
+import { catchError, firstValueFrom, map } from 'rxjs';
+// eslint-disable-next-line boundaries/element-types
```
Yes, I wanted to talk to you about this today but then I realized you wouldn't be there in person. I thought it would be elegant to maintain the Component instance in the service which obviously breaks the rule but I also think it makes sense in this case. We can add a rule like that but I'm not sure if that's not a bit over-engineered as long as it's just the one case. But I'm open to either solution.

Cito replied:
Let's discuss it on Monday, I already have an idea.

SilverLinings89 commented on src/app/data-access/services/data-access.service.ts:
```diff
@@ -0,0 +1,180 @@
+  #arsEndpointUrl = `${this.#arsBaseUrl}/access-requests`;
+
+  showNewAccessRequestDialog = (datasetID: string) => {
+    const user = this.#auth.user();
+
+    if (!this.#auth.isAuthenticated() || !user) {
```
I try to write code in situations like this with as few assumptions as possible, because that fact may change in the future and then break this code if it does so I thought I would check everything I'll need here. However, since it's not likely that this will change soon I'll just remove it for simplicity of the code.

Cito commented on src/app/access-requests/features/data-access-request-dialog/data-access-request-dialog.component.ts:
```diff
@@ -0,0 +1,132 @@
+    FormsModule,
+    MatDialogContent,
+    MatDialogModule,
+    MatDialogTitle,
```
Actually just noticed that all the MatDialog... components are contained in the MatDialogModule already, so you only need to import that. Can be also changed elsewhere.

SilverLinings89 replied:
Will do this separately

Cito commented on src/app/access-requests/features/data-access-request-dialog/data-access-request-dialog.component.ts:
```diff
@@ -0,0 +1,132 @@
+  };
+
+  cancelClick = () => {
+    this.data.isCanceled = true;
```
you don't need the isCanceled property - the data will not be returned in this case anyway

Cito replied:
currently if you click outside the dialog, it simply closes. this is not nice if the form is dirty (the user already has entered data). should we somehow disallow this or add a confirmation dialog?

SilverLinings89 replied:
I will remove it separately. I don't think we need the separate confirmation. That just bloats it and I believe it is a very rare edge case

SilverLinings89 replied:
done.

Cito commented on src/app/data-access/features/granted-access-requests-list/granted-access-requests-list.component.html:
```diff
@@ -0,0 +1,20 @@
+      }
+    </ul>
+  }
+}
```
Looks like the button is still missing.

Cito reviewed (approved):
There are still some unresolved issues. If you want to resolved them in a follow-up PR feel free to merge already. These should just not be forgotten.

SilverLinings89 commented on src/app/access-requests/features/data-access-request-dialog/data-access-request-dialog.component.html:
```diff
@@ -0,0 +1,61 @@
+        [min]="minFromDate"
+        [max]="maxFromDate"
+        [(ngModel)]="fromDate"
+        (dateInput)="fromDateChanged($event)"
```
yes

SilverLinings89 commented on src/app/access-requests/features/data-access-request-dialog/data-access-request-dialog.component.html:
```diff
@@ -0,0 +1,61 @@
+        matInput
+        placeholder="your.email@goes.here"
+        #emailField
+        [(ngModel)]="email"
```
I will work on this more in the future as a separate PR to validate the form

SilverLinings89 commented on src/app/access-requests/features/data-access-request-dialog/data-access-request-dialog.component.html:
```diff
@@ -0,0 +1,61 @@
+  </mat-dialog-content>
+  <mat-dialog-actions>
+    <button mat-stroked-button class="mr-2" (click)="cancelClick()">Cancel</button>
+    <button mat-raised-button (click)="submitClick()">Submit</button>
```
see below

SilverLinings89 commented on src/app/access-requests/features/data-access-request-dialog/data-access-request-dialog.component.html:
```diff
@@ -0,0 +1,61 @@
+        #details
+      ></textarea>
+    </mat-form-field>
+    min: {{ minFromDate }} + {{ maxFromDate }}
```
argh. sorry.

SilverLinings89 commented:
I fixed the urgent things only (console.log, debugging output in the ui and removed the isCanceled property from the return type. The rest can be a separate improvement.
